File: src/jastudio/jastudio/language_services/janome_ex/word_extraction/matches/state_tests/head/requires_or_forbids_generic.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from jastudio.language_services.janome_ex.word_extraction.matches.state_tests.head.failed_match_requirement import FailedMatchRequirement

if TYPE_CHECKING:
    from collections.abc import Callable

    from jastudio.language_services.janome_ex.word_extraction.matches.requirements.vocab_match_inspector import VocabMatchInspector
    from jastudio.note.notefields.require_forbid_flag_field import RequireForbidFlagField

class RequiresOrForbids(Slots):
    def __init__(self, name: str,
                 get_requirement: Callable[[VocabMatchInspector], RequireForbidFlagField],
                 is_in_state: Callable[[VocabMatchInspector], bool]) -> None:
        self._get_requirement: Callable[[VocabMatchInspector], RequireForbidFlagField] = get_requirement
        self._is_in_state: Callable[[VocabMatchInspector], bool] = is_in_state
        self._required_failure: FailedMatchRequirement = FailedMatchRequirement.required(name)
        self._forbidden_failure: FailedMatchRequirement = FailedMatchRequirement.forbids(name)

    def apply_to(self, inspector: VocabMatchInspector) -> FailedMatchRequirement | None:
        requirement = self._get_requirement(inspector)

        if requirement.is_required and not self._is_in_state(inspector):
            return self._required_failure
        if requirement.is_forbidden and self._is_in_state(inspector):
            return self._forbidden_failure
        return None

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/WordExtraction/Matches/StateTests/Head/RequiresOrForbidsGeneric.cs
using System;
using JAStudio.Core.LanguageServices.JanomeEx.WordExtraction.Matches.Requirements;
using JAStudio.Core.Note.NoteFields;

namespace JAStudio.Core.LanguageServices.JanomeEx.WordExtraction.Matches.StateTests.Head;

sealed class RequiresOrForbids
{
   readonly Func<VocabMatchInspector, RequireForbidFlagField> _getRequirement;
   readonly Func<VocabMatchInspector, bool> _isInState;
   readonly FailedMatchRequirement _requiredFailure;
   readonly FailedMatchRequirement _forbiddenFailure;

   public RequiresOrForbids(
      string name,
      Func<VocabMatchInspector, RequireForbidFlagField> getRequirement,
      Func<VocabMatchInspector, bool> isInState)
   {
      _getRequirement = getRequirement;
      _isInState = isInState;
      _requiredFailure = FailedMatchRequirement.Required(name);
      _forbiddenFailure = FailedMatchRequirement.Forbids(name);
   }

   public FailedMatchRequirement? ApplyTo(VocabMatchInspector inspector)
   {
      var requirement = _getRequirement(inspector);

      if(requirement.IsRequired && !_isInState(inspector))
      {
         return _requiredFailure;
      }

      if(requirement.IsForbidden && _isInState(inspector))
      {
         return _forbiddenFailure;
      }

      return null;
   }
}
